Sources/VGSShowSDK/Utils/Extensions/Foundation+Extensions/String+Exntesions.swift

```diff
@@ -197,3 +197,25 @@ internal extension String {
 		return path
 	}
 }
+internal extension String {
+  
+  func secure(in range: VGSTextRange, secureSymbol: String? = "*") -> String {
+    guard !self.isEmpty else {
+      return ""
+    }
+    
+    let securedString: String
+    if range.start == nil, range.end == nil {
+      securedString = String(describing: Array(repeating: secureSymbol, count: self.count))
+    } else {
+      let first = range.start ?? 0
+      let last = range.end ?? self.count - 1
+      let startIndex = self.index(self.startIndex, offsetBy: first)
```

Comment:
This func should be rewritten:
1. `startIndex` and `endIndex` are default Swift String properties.
https://developer.apple.com/documentation/swift/string/1540930-startindex
https://developer.apple.com/documentation/swift/array/1539310-endindex

Shadowing them is not a good idea.

2. There are no checks for start/end boundaries. 

Code with wrong range will crash.
```
func testSecureString() {
	let text = "2"
	let secureText = text.secure(in: (0, 4))
	print(secureText)
}

testEmptyString()
```

Will try to update this later.

Reply:
```swift
typealias VGSTextRange = (start: Int?, end: Int?)

extension String {

	/// First secure index from secure range. `0` if start is less than `0` or `nil`.
	/// - Parameter range: `VGSTextRange` object.
	/// - Returns: `Int` object, index of start position.
	func firstSecureIndex(from range: VGSTextRange) -> Int {
		let first: Int
		if let start = range.start, start >= 0 {
			first = start
		} else {
			first = 0
		}

		return first
	}

	/// Last secure index from secure range. `0` if `nil` or empty string.
	/// - Parameter range: `VGSTextRange` object.
	/// - Returns: `Int` object, index of start position.
	func lastSecureIndex(from range: VGSTextRange) -> Int {
		let end: Int
		if let last = range.end, last >= 0 {
			end = last
		} else {
			if !isEmpty {
				end = count - 1
			} else {
				end = 0
			}
		}

		return end
	}

	/// Characters count to replace.
	/// - Parameter range: `VGSTextRange` object.
	/// - Returns: `Int` object. Characters count to replace.
	func charactersCountToReplace(from range: VGSTextRange) -> Int {
		let first = firstSecureIndex(from: range)
		let last = lastSecureIndex(from: range)

		return last - first
	}

	/// Convert int range to Swift String index range. If start of end are beyond the string boundaries, default `startIndex` and `endIndex` will be used.
	/// - Parameter range: `VGSTextRange` object.
	/// - Returns: `Range<String.Index>` object.
	func stringRangeFromIntRange(range: VGSTextRange) -> Range<String.Index> {

		let first = firstSecureIndex(from: range)
		let last = lastSecureIndex(from: range)

		let replacementStartIndex: Index
		let replacementEndIndex: Index

		if let index = self.index(self.startIndex, offsetBy: first, limitedBy: endIndex) {
			replacementStartIndex = index
		} else {
			replacementStartIndex = startIndex
		}

		if let index = self.index(self.startIndex, offsetBy: last, limitedBy: endIndex) {
			replacementEndIndex = index
		} else {
			replacementEndIndex = endIndex
		}

		return replacementStartIndex..<replacementEndIndex
	}

	/// Return replacement string from secure range and secureSymbol.
	/// - Parameters:
	///   - range: `VGSTextRange` object.
	///   - secureSymbol: `String` object, default is `*`.
	/// - Returns: `String` object. Replacement secured string.
	func replacementStringFromSecureRange(_ range: VGSTextRange, secureSymbol: String = "*") -> String {
		let first = firstSecureIndex(from: range)
		let last = lastSecureIndex(from: range)

		let replaceCount = last - first

		/// Replace all if negative valur or range is bigger that string length.
		if replaceCount <= 0 || replaceCount > count {
			return String(repeating: secureSymbol, count: count)
		}

		return String(repeating: secureSymbol, count: replaceCount)
	}

	/// Return new secured string from range and secureSymbol.
	/// - Parameters:
	///   - range: `VGSTextRange` object.
	///   - secureSymbol: `String` object, default is `*`.
	/// - Returns: `String` object, secured string.
	func secure(in range: VGSTextRange, secureSymbol: String = "*") -> String {
		guard !self.isEmpty else {
			return ""
		}

		let securedString: String
		if range.start == nil, range.end == nil {
			securedString = String(describing: Array(repeating: secureSymbol, count: count))
		} else {
			let replacementRange = stringRangeFromIntRange(range: range)
			let replacementString = replacementStringFromSecureRange(range, secureSymbol: secureSymbol)
			securedString = self.replacingCharacters(in: replacementRange, with: replacementString)
		}
		return securedString
	}
}

func testSecureString1() {
	let text = "2"
	let secureText = text.secure(in: (0, 4))
	print(secureText)
}

func testSecureString2() {
	let text = "4111-1111-1111-1111"
	let secureText = text.secure(in: (0, 4))
	print(secureText)
}

func testSecureString3() {
	let text = "4111-1111-1111-1111"
	let secureText = text.secure(in: (-1, 4))
	print(secureText)
}

func testSecureString4() {
	let text = "4111-1111-1111-1111"
	let secureText = text.secure(in: (-1, 4))
	print(secureText)
}

func testSecureString5() {
	let text = "4111-1111-1111-1111"
	let secureText = text.secure(in: (-1, 4), secureSymbol: "🐒")
	print(secureText)
}

func testSecureString6() {
	let text = "4111-1111-1111-1111"
	let secureText = text.secure(in: (60, 50), secureSymbol: "*")
	print(secureText)
}

func testSecureString7() {
	let text = "4111-1111-1111-1111"
	let secureText = text.secure(in: (5, 9), secureSymbol: "*")
	print(secureText)
}

func testSecureString8() {
	let text = ""
	let secureText = text.secure(in: (0, 4))
	print(secureText)
}

testSecureString1()
testSecureString2()
testSecureString3()
testSecureString4()
testSecureString5()
testSecureString6()
testSecureString7()
testSecureString8()

```